### PR TITLE
Reset job disabling timer on adding the job again

### LINKED ISSUE #30285: Background job execution delayed by 12 hours during app reenabling
Steps to reproduce:
- Have a job in an app that is scheduled
- Disable the app
- The job gets executed
 - Class cannot be found in https://github.com/nextcloud/server/blob/7239b3d560b52ed604484a0a549dd5805e3f21e4/lib/private/BackgroundJob/JobList.php#L282
 - The job gets delayed for the next 12 hours https://github.com/nextcloud/server/blob/7239b3d560b52ed604484a0a549dd5805e3f21e4/lib/private/BackgroundJob/JobList.php#L216-L222
- Reenable the app

Current behavior: 
The job will run earliest after 12 hours

Expected behavior:
The job last checked date is reset so that it runs immediately again once the app is enabled

https://github.com/nextcloud/server/blob/06611e47802661b5ffbdb62c8c9d630083fd3879/lib/private/legacy/OC_App.php#L1074
https://github.com/nextcloud/server/blob/7239b3d560b52ed604484a0a549dd5805e3f21e4/lib/private/BackgroundJob/JobList.php#L68

cc @nextcloud/server-backend

### PR DESCRIPTION
**Pull request by juliusknorr**
If a job run is triggered while having the app disabled the next execution will be delayed for 12 hours 

https://github.com/nextcloud/server/blob/60c3f15fa40840648d8fbb0192cfbd956ff1fe8a/lib/private/BackgroundJob/JobList.php#L235-L241

This PR ensures that the job is rescheduled immediately if an app gets enabled and the jobs from appinfo.xml get parsed again in https://github.com/nextcloud/server/blob/37f40cdd468a2826f6269bacc81ff74d8c606d23/lib/private/Installer.php#L160

Fixes https://github.com/nextcloud/server/issues/30285

### CODE REVIEW

File: lib/private/BackgroundJob/JobList.php [hunk 1]
```diff
@@ -65,6 +65,7 @@ public function __construct(IDBConnection $connection, IConfig $config, ITimeFac
 	 * @param mixed $argument
 	 */
 	public function add($job, $argument = null) {
+		$query = $this->connection->getQueryBuilder();
```
Comment (skjnldsv) [inline lib/private/BackgroundJob/JobList.php:68]: Moving was most likely not necessary since you ended up using `$reset`, but all good :)

**Review by skjnldsv — APPROVED** [review]

File: lib/private/BackgroundJob/JobList.php (hunk 1 above)
Comment (juliushaertl) [reply lib/private/BackgroundJob/JobList.php]: Addressed, lets keep those separated then

**Comment by artonge** [comment]
Can we prevent duplicate code? But maybe I missed some tiny difference :).

```php
	public function add($job, $argument = null) {
		if ($job instanceof IJob) {
			$class = get_class($job);
		} else {
			$class = $job;
		}

		$argument = json_encode($argument);
		if (strlen($argument) > 4000) {
			throw new \InvalidArgumentException('Background job arguments can\'t exceed 4000 characters (json encoded)');
		}

		$query = $this->connection->getQueryBuilder();

		if (!$this->has($job, $argument)) {
			$query->insert('jobs')
				->values([
					'class' => $query->createNamedParameter($class),
					'argument' => $query->createNamedParameter($argument),
					'last_run' => $query->createNamedParameter(0, IQueryBuilder::PARAM_INT),
					'last_checked' => $query->createNamedParameter($this->timeFactory->getTime(), IQueryBuilder::PARAM_INT),
				]);
		} else {
			$query->update('jobs')
				->set('reserved_at', $query->expr()->literal(0, IQueryBuilder::PARAM_INT))
				->set('last_checked', $query->createNamedParameter($this->timeFactory->getTime(), IQueryBuilder::PARAM_INT))
				->where($query->expr()->eq('class', $query->createNamedParameter($class)))
				->andWhere($query->expr()->eq('argument', $query->createNamedParameter($argument)));
		}

		$query->execute();
	}
```

**Comment by juliushaertl** [comment]
@artonge Pushed a fixup with some minor adjustment because `$this->has($job, $argument)` requires the original argument instead of the json_encoded one.

**Review by artonge — APPROVED** [review]
We can move `$query->execute();` out of the `if`, but ok for me :)

**Comment by juliushaertl** [comment]
> We can move $query->execute(); out of the if, but ok for me :)

Rebased, squashed, replaced deprecated execute with executeStatement and moved it out of the if as well ;)

**Pull request merged**